src/client/eppo-client.ts

```diff
@@ -82,168 +82,174 @@ export default class EppoClient implements IEppoClient {
     subjectKey: string,
-    experimentKey: string,
+    flagKey: string,
     subjectAttributes = {},
     assignmentHooks: IAssignmentHooks | undefined,
     valueType: ValueType,
-  ): EppoValue | null {
+  ): { allocationKey: string | null, assignment: EppoValue | null } {
     validateNotBlank(subjectKey, 'Invalid argument: subjectKey cannot be blank');
-    validateNotBlank(experimentKey, 'Invalid argument: experimentKey cannot be blank');
+    validateNotBlank(flagKey, 'Invalid argument: flagKey cannot be blank');
+
+    const nullAssignment = { allocationKey: null, assignment: null };
 
-    const experimentConfig = this.configurationStore.get<IExperimentConfiguration>(experimentKey);
+    const experimentConfig = this.configurationStore.get<IExperimentConfiguration>(flagKey);
     const allowListOverride = this.getSubjectVariationOverride(
       subjectKey,
       experimentConfig,
       valueType,
     );
 
-    if (allowListOverride) return allowListOverride;
+    if (allowListOverride) return { ...nullAssignment, assignment: allowListOverride };
+
 
     // Check for disabled flag.
-    if (!experimentConfig?.enabled) return null;
+    if (!experimentConfig?.enabled) return nullAssignment;
 
     // check for overridden assignment via hook
-    const overriddenAssignment = assignmentHooks?.onPreAssignment(experimentKey, subjectKey);
+    const overriddenAssignment = assignmentHooks?.onPreAssignment(flagKey, subjectKey);
     if (overriddenAssignment !== null && overriddenAssignment !== undefined) {
-      return overriddenAssignment;
+      return { ...nullAssignment, assignment: overriddenAssignment };
     }
 
     // Attempt to match a rule from the list.
     const matchedRule = findMatchingRule(subjectAttributes || {}, experimentConfig.rules);
-    if (!matchedRule) return null;
+    if (!matchedRule) return nullAssignment;
 
     // Check if subject is in allocation sample.
     const allocation = experimentConfig.allocations[matchedRule.allocationKey];
-    if (!this.isInExperimentSample(subjectKey, experimentKey, experimentConfig, allocation))
-      return null;
+    if (!this.isInExperimentSample(subjectKey, flagKey, experimentConfig, allocation))
+      return nullAssignment;
 
     // Compute variation for subject.
     const { subjectShards } = experimentConfig;
     const { variations } = allocation;
 
-    const shard = getShard(`assignment-${subjectKey}-${experimentKey}`, subjectShards);
+    const shard = getShard(`assignment-${subjectKey}-${flagKey}`, subjectShards);
     const assignedVariation = variations.find((variation) =>
       isShardInRange(shard, variation.shardRange),
     )?.typedValue;
 
+    const allocationKey = { allocationKey: matchedRule.allocationKey };
+
     switch (valueType) {
       case ValueType.BoolType:
-        return EppoValue.Bool(assignedVariation as boolean);
+        return { ...allocationKey, assignment: EppoValue.Bool(assignedVariation as boolean) };
```

Comment:
Not your code, but I believe by telling TypeScript to treat these things as a type, we could actually end up passing in a numeric and will just be a number stored (and returned from) `boolValue`. It only matters in the edge case of a developer requesting the incorrect variant type for an experiment. But this is a different behavior than our PHP SDK (which would return null) and Java SDK (which would cast)

Reply:
I agree that this behavior is too divergent from the other SDKs and needs more guardrails. I just added a `isExpectedType` method to `EppoValue` like I did in the python and ruby SDKs and return `NullType` if there's a mismatch. I do a `typeof <numeric_variable> === 'number'` comparison. That should suffice, right?